dali/operators/random/rng_base_gpu.cuh

```diff
@@ -92,22 +139,49 @@ void RNGBase<Backend, Impl, IsNoiseGen>::RunImplTyped(workspace_t<GPUBackend> &w
     in_view = view<const T>(input);
     output.SetLayout(input.GetLayout());
   }
-  blockdesc_count = SetupBlockDescs(
-    blocks_cpu, block_sz, max_nblocks, out_view, in_view);
+
+  // TODO(janton): set layout explicitly from the user for RNG
+  auto layout = output.GetLayout();
+  bool independent_channels = This().PerChannel();
+
+  // Channel dimension is specified only when we follow the "generate once, apply to
+  // all channels" aproach
+  int channel_dim = -1;
+  if (!independent_channels) {
+    channel_dim = layout.empty() ? out_view.shape.sample_dim() - 1 : layout.find('C');
+  }
+
+  auto &samples_cpu = backend_data_.sample_descs_cpu_;
+  auto &samples_gpu = backend_data_.sample_descs_gpu_;
+  int nsamples = SetupSampleDescs(samples_cpu.data(), out_view, in_view, channel_dim);
+  if (nsamples == 0) {
+    return;
+  }
+  samples_gpu.from_host(samples_cpu, ws.stream());
+
+  auto &blocks_cpu = backend_data_.block_descs_cpu_;
+  auto &blocks_gpu = backend_data_.block_descs_gpu_;
+  auto shape_copy = out_view.shape;
+  if (!independent_channels) {
+    // For generate once, apply to all channels, the channel dimension will be
+    // removed, so that each CUDA thread always processes all channels of a pixel.
+    for (int s = 0; s < nsamples; s++) {
+      shape_copy.tensor_shape_span(s)[channel_dim] = 1;
```

Comment:
Won't this change affect `sample_siz` inside `SetupBlockDescs`?

Reply:
yes, that's the whole point. In this case we calculate blocks as number of full pixels. SetupBlockDesc is dividing a shape into blocks. Here we remove the channel dimension from the shape copy, calculate blocks, and then launch a different kernel, that uses c_count and c_stride to visit all channels in every point.

Reply:
ok

Reply:
I think id would better to pass the original shape to SetupBlockDescs and just skip `channel_dim` when computing the volume.

Reply:
Will do